internal/status/status.go

```diff
@@ -0,0 +1,318 @@
+)
+
+//go:generate mockgen -source=status.go -package=status -destination=mock_status.go StatusAPI
+
+type StatusAPI interface {
+	GetConditions(ctx context.Context, nfdInstance *nfdv1.NodeFeatureDiscovery) []metav1.Condition
+	AreConditionsEqual(prevConditions, newConditions []metav1.Condition) bool
+}
+
+type status struct {
+	helper statusHelperAPI
+}
+
+func NewStatusAPI(deploymentAPI deployment.DeploymentAPI, daemonsetAPI daemonset.DaemonsetAPI) StatusAPI {
+	helper := newStatusHelperAPI(deploymentAPI, daemonsetAPI)
+	return &status{
+		helper: helper,
+	}
+}
+
+func (s *status) GetConditions(ctx context.Context, nfdInstance *nfdv1.NodeFeatureDiscovery) []metav1.Condition {
+	// get worker daemonset conditions
+	nonAvailableConditions := s.helper.getDaemonSetNotAvailableConditions(ctx, nfdInstance.Namespace, "nfd-worker")
+	if nonAvailableConditions != nil {
+		return nonAvailableConditions
+	}
+	// get master deployment conditions
+	nonAvailableConditions = s.helper.getDeploymentNotAvailableConditions(ctx, nfdInstance.Namespace, "nfd-master")
+	if nonAvailableConditions != nil {
+		return nonAvailableConditions
+	}
+	// get GC deployment conditions
+	nonAvailableConditions = s.helper.getDeploymentNotAvailableConditions(ctx, nfdInstance.Namespace, "nfd-gc")
+	if nonAvailableConditions != nil {
+		return nonAvailableConditions
+	}
+	// get topology, if needed
+	if nfdInstance.Spec.TopologyUpdater {
+		nonAvailableConditions := s.helper.getDaemonSetNotAvailableConditions(ctx, nfdInstance.Namespace, "nfd-topology-updater")
+		if nonAvailableConditions != nil {
+			return nonAvailableConditions
+		}
+	}
+
+	return getAvailableConditions()
+}
+
+func (s *status) AreConditionsEqual(prevConditions, newConditions []metav1.Condition) bool {
+	for _, newCondition := range newConditions {
+		oldCondition := meta.FindStatusCondition(prevConditions, newCondition.Type)
+		if oldCondition == nil {
+			return false
+		}
+		// Ignore timestamps
+		if oldCondition.Status != newCondition.Status ||
+			oldCondition.Reason != newCondition.Reason ||
+			oldCondition.Message != newCondition.Message {
+			return false
+		}
+	}
+	return true
+}
+
+//go:generate mockgen -source=status.go -package=status -destination=mock_status.go statusHelperAPI
+
+type statusHelperAPI interface {
+	getDaemonSetNotAvailableConditions(ctx context.Context, namespace, name string) []metav1.Condition
+	getDeploymentNotAvailableConditions(ctx context.Context, namespace, name string) []metav1.Condition
+}
+
+type statusHelper struct {
+	deploymentAPI deployment.DeploymentAPI
+	daemonsetAPI  daemonset.DaemonsetAPI
+}
+
+func newStatusHelperAPI(deploymentAPI deployment.DeploymentAPI, daemonsetAPI daemonset.DaemonsetAPI) statusHelperAPI {
+	return &statusHelper{
+		deploymentAPI: deploymentAPI,
+		daemonsetAPI:  daemonsetAPI,
+	}
+}
+
+func (sh *statusHelper) getDaemonSetNotAvailableConditions(ctx context.Context, dsNamespace, dsName string) []metav1.Condition {
```

Comment:
nit: Maybe rename this to indicate that this is solely for the nfd-worker daemonset. Or alternatively make the function more generic?

```suggestion
func (sh *statusHelper) getWorkerDaemonSetNotAvailableConditions(ctx context.Context, dsNamespace, dsName string) []metav1.Condition {
```

Reply:
This function is generic (it deals with both worker and topolgy daemonsets), but i will re-write it to make the code clearer